nabs/suspenders.py

```diff
@@ -0,0 +1,179 @@
+        suspend state. If suspended but we should_resume, we'll start a timer
+        of length sleep to clear the suspend state. This will be interrupted if
+        another value change leads us to should_suspend.
+        """
+        with self._rlock:
+            if self._suspend_ev.is_set():
+                if self.should_resume(value):
+                    self._suspend_ev.clear()
+                    self._run_release_thread()
+            else:
+                if self.should_suspend(value):
+                    logger.info('Suspending due to bad %s value=%s',
+                                self._sig.name, value)
+                    loop = asyncio.get_event_loop()
+                    self._ok_future = loop.create_future()
+                    self._suspend_ev.set()
+
+    def _run_release_thread(self):
+        logger.info('%s suspension is over, waiting for %ss then resuming.',
+                    self._sig.name, self._sleep)
+        loop = asyncio.get_event_loop()
+        loop.run_in_executor(None, self._release_thread)
+
+    def _release_thread(self):
+        logger.debug('Worker waiting to release suspender...')
+        if self._suspend_ev.wait(timeout=self._sleep):
+            logger.debug('Worker canceling suspender release')
+        else:
+            logger.debug('Worker releasing suspender')
+            self._ok_future.set_result('ok')
+
+    def __call__(self, plan):
+        """
+        Mutate plan to call self._msg_proc on each msg that comes through.
+        """
+        logger.debug('Running plan with suspender')
+        if self._subid is None:
+            self._subid = self._sig.subscribe(self._update,
+                                              event_type=self._sig.SUB_VALUE)
+        try:
+            yield from plan_mutator(plan, self._msg_proc)
+        finally:
+            if self._subid is not None:
+                self._sig.unsubscribe(self._subid)
+                self._subid = None
+
+    def _msg_proc(self, msg):
+        """
+        At each msg, decide if we should wait for a suspension to lift.
+        """
+        logger.debug('enter msg_proc')
+        with self._rlock:
+            if self._cmd is None or msg.command in self._cmd:
+                if not self._ok_future.done() and not self._suspend_active:
+                    return self._suspend(msg), None
+            return None, None
+
+    def _suspend(self, msg):
+        logger.debug('suspend on msg=%s', msg)
+        # Set this flag so we don't recursively check to suspend on wait_for
+        self._suspend_active = True
+        yield from wait_for([self._ok_future])
+        self._suspend_active = False
+        logger.debug('resuming plan in _suspend')
+        # Yield a copy of the message so we can suspend again
+        # Bluesky skips preprocessor if it already saw the message
+        return (yield copy.copy(msg))
+
+
+class BeamSuspender(SuspendPreprocessor):
```

Comment:
This is awesome! Very clear and elegant. Means you did nice work in the class design 😉.

On a completely unrelated note it makes me want a global `BeamStats`.

Reply:
Agreed on global `BeamStats`. This can be revisited.

Reply:
One caveat: we will have two beams in a year or more.